sql/expression/in.go

```diff
@@ -240,11 +251,75 @@ func newInMap(expr sql.Expression) (map[uint64]sql.Expression, bool, sql.Type, e
 	return elements, hasNull, t, nil
 }
 
-// copied variant from cache.go
-func hashOf(i interface{}) (uint64, error) {
+func hashOf(e sql.Expression, t sql.Type) (uint64, error) {
+	switch v := e.(type) {
+	case Tuple:
+		return hashOfTuple(v)
+	case *Literal:
+		return hashOfLiteral(v, t)
+	default:
+		return 0, ErrUnsupportedHashInSubexpression.New(v)
+	}
+}
+
+func hashOfLiteral(l *Literal, t sql.Type) (uint64, error) {
 	hash := xxhash.New()
+	i, err := t.Convert(l.value)
+	if err != nil {
+		return 0, err
+	}
 	if _, err := hash.Write([]byte(fmt.Sprintf("%#v,", i))); err != nil {
 		return 0, err
 	}
 	return hash.Sum64(), nil
 }
+
+// hashOfTuple will recursively hash a Tuple tree with Literal leaves
+func hashOfTuple(tup Tuple) (uint64, error) {
+	hash := xxhash.New()
+	for _, el := range tup {
+		switch v := el.(type) {
+		case *Literal:
+			if _, err := hash.Write([]byte(fmt.Sprintf("%#v,", v.value))); err != nil {
+				return 0, err
+			}
+		case Tuple:
```

Comment:
This is probably more trouble that it's worth. The case we care about is `(col_a, col_b) in ( (1,2), (3,4) )`

Would rip this out and not do this transformation if there is more than a single level of nested tuples. Then see if this can be simplified at all.

Reply:
replaces the recursion line with an error, so not a lot shorter, but it might let me do the tuple element casting. will try to add that before merging